Fix use arm neon instructions if targeting 32-bit arm architectures only

gilzoide opened the pull request:
This fixes compilation for arm64 platforms that support neon instructions, like Apple Silicon, because the assembly code contained in `pixman-arm-neon-asm.S` is armv7 assembly, which is not supported by arm64 compilers.

Fixes #496.

gilzoide committed:
Fix use arm neon instructions if targeting 32-bit arm architectures only

This fixes compilation for arm64 platforms, like Apple Silicon, because the assembly code contained in `pixman-arm-neon-asm.S` is armv7 assembly, which is not supported by arm64 compilers.

Fe1777 commented:
The provided issue description suggests that there's an oversight or bug related to the inclusion of ARM NEON instructions when targeting 32-bit ARM architectures in the Samsung/rlottie repository. NEON is an advanced SIMD (Single Instruction Multiple Data) architecture extension for ARM processors, which can significantly enhance performance for multimedia and signal processing tasks.

### **Fixing ARM NEON Inclusion in rlottie**

To ensure that ARM NEON instructions are only included when targeting 32-bit ARM architectures, you need to update the build configuration, likely modifying cmake files or conditional preprocessor directives to properly detect and handle ARM architecture targets.

Here's a step-by-step solution assuming the usage of CMake for build configuration:

---

### **Step 1: Detect ARM and NEON Support**

Ensure that the build system accurately detects whether it is targeting a 32-bit ARM architecture and if the NEON instruction set is available.

#### Update `CMakeLists.txt`

1. **Detecting ARM Architecture:**
   Use CMake to detect if the build is for ARM and if it's a 32-bit architecture.

2. **Checking for NEON Support:**
   Check if NEON is supported on the detected ARM architecture.

```cmake
# Add this to your CMakeLists.txt file or a relevant cmake configuration file

include(CheckCCompilerFlag)

function(check_arm_neon)
    if(CMAKE_SYSTEM_PROCESSOR MATCHES "arm" OR CMAKE_SYSTEM_PROCESSOR MATCHES "ARM")
        message(STATUS "Targeting ARM architecture")

        # Check for 32-bit ARM
        if(CMAKE_SIZEOF_VOID_P EQUAL 4)
            message(STATUS "32-bit ARM detected")
            set(ARMV7 TRUE)
        endif()

        # Check for NEON support
        check_c_compiler_flag("-mfpu=neon" COMPILER_HAS_NEON)
        if(COMPILER_HAS_NEON)
            message(STATUS "NEON supported")
            set(ARM_NEON_SUPPORTED TRUE)
        else()
            message(STATUS "NEON not supported")
            set(ARM_NEON_SUPPORTED FALSE)
        endif()
    endif()
endfunction()

check_arm_neon()
```

### **Step 2: Conditional Compilation Based on Detection**

Modify the relevant source files to conditionally include NEON-specific code only when ARM NEON support is detected.

#### Update Source Files with Preprocessor Directives

In your code, wrap the NEON-specific code sections with appropriate preprocessor directives:

```c
#ifdef ARM_NEON_SUPPORTED
#include <arm_neon.h>
#endif

// Example function utilizing NEON
void example_function() {
#ifdef ARM_NEON_SUPPORTED
    // NEON optimized code
    float32x4_t a = vld1q_f32(input);
    float32x4_t b = vdupq_n_f32(2.0);
    float32x4_t result = vmulq_f32(a, b);
    vst1q_f32(output, result);
#else
    // Fallback non-NEON code
    for (int i = 0; i < 4; i++) {
        output[i] = input[i] * 2.0;
    }
#endif
}
```

### **Step 3: Update Build Instructions**

Ensure that the build system includes the correct compiler flags when NEON is supported:

#### Update `CMakeLists.txt` to Append Flags

```cmake
if(ARMV7 AND ARM_NEON_SUPPORTED)
    add_compile_options(-mfpu=neon)
    message(STATUS "Added NEON compile flag")
endif()
```

### **Concluding the Solution**

After following these steps:

1. The `CMakeLists.txt` will correctly detect if the target architecture is 32-bit ARM and if NEON is supported.
2. Conditional compilation will ensure that NEON-specific code is only used when appropriate.
3. Compiler flags will be updated to enable NEON instructions when detected.

### **Testing and Validation**

1. **Build Verification:**
   - Build the project targeting different architectures, including both 32-bit ARM with and without NEON support, and verify the build output.

2. **Runtime Verification:**
   - Execute the binaries on corresponding hardware to ensure the NEON optimizations take effect and fallbacks are correctly used where NEON is not supported.

---

By making these changes, you should be able to fix the inclusion of ARM NEON instructions to be specific to 32-bit ARM architectures, ensuring optimal performance without affecting compatibility. Make sure to thoroughly test the changes across all targeted platforms.

Fe1777 commented:
Just checking... Easy...

Fe1777 commented:
> Hey @Fe1777, this is a PR that already fixes the problem in question, not an open issue in need of "how to fix" instructions generated by AI.

No wonder you going to be fired...
So sad...

gilzoide commented:
> No wonder you going to be fired...
> So sad...

This makes no sense at all in this context.
@Fe1777 what are you trying to accomplish here?

gilzoide committed:
Fix don't include pixman arm neon code when compiling with clang

Newer versions of the Android NDK use clang for compiling, which cannot compile the code contained in `pixman-arm-neon-asm.S`.

gilzoide commented:
Now this PR also fixes compilation when using clang, for example for build rlottie for Android + NDK.

thomastay commented:
This worked for me on mac, would be great if it could be merged

mdingwall commented:
To add fuel to this, I have successfully used this to build on MacOS 15.1.1 on an M1 Pro MacBook Pro and in an AARCH64 Ubuntu 22.04 VM on the same machine.
